src/expr/src/lib.rs

```diff
@@ -61,14 +62,16 @@ impl OptimizedMirRelationExpr {
     }
 }
 
-impl AsRef<MirRelationExpr> for OptimizedMirRelationExpr {
-    fn as_ref(&self) -> &MirRelationExpr {
+impl Deref for OptimizedMirRelationExpr {
+    type Target = MirRelationExpr;
+
+    fn deref(&self) -> &Self::Target {
         &self.0
     }
 }
 
-impl AsMut<MirRelationExpr> for OptimizedMirRelationExpr {
-    fn as_mut(&mut self) -> &mut MirRelationExpr {
+impl DerefMut for OptimizedMirRelationExpr {
+    fn deref_mut(&mut self) -> &mut Self::Target {
         &mut self.0
     }
 }
```

Comment:
I think this is a point of contention, in that the wrapper is meant to "ensure" that the type stays optimized, and it should be a code smell to call `as_mut()`. The `Deref` implementation seems like no worries, but .. I could see disagreement on this one.

Reply:
This is a good point. In my opinion none of the methods convey the intent as the contract of both AsRef and DerefMut is defined by std.

If the caller needs to maintain some extra invariant, like not mutating an optimized expression into a non-optimized expression, I would add a method directly on the struct with documentation to that effect. I pushed a version of this in this PR.

If you'd rather keep the `as_mut` for the mutable access I'm happy to put it back

Reply:
Nope, this seems great to me.